src/core/toga/window.py

```diff
@@ -181,9 +186,19 @@ def full_screen(self, is_full_screen):
     def close(self):
         self._impl.close()
 
-    def on_close(self):
+    def toga_on_close(self):
+        self.app.windows -= self
         self._impl.on_close()
```

Comment:
I can see what you're doing here; however, the old "on_close()" method was probably a mistake. By having it on the interface layer, we're effectively saying "this is a method you *could* invoke" - which definitely isn't true. 

Since all we're ever going to be doing here is:
1. removing the window from the `self.windows` list, and
2. triggering the *user's* `on_close` handler
we may be better off leaving this logic 100% on the implementation side. Yes, this does mean we're going to be duplicating the "remove window, invoke hander" logic on every backend... but then, that's already true for every other event handler anyway, and it allows us to customise exactly how and when the events are triggered (in the unlikely event that ordering ever matters).

Reply:
Why not simply making it private with the "__" prefix? isn't it enough in order to indicate that this is not a method that should be called?

Reply:
I transferred the `on_close` logic to implementation side, basically copying `button`'s `on_press` handler behavior... I've tested it on Windows, and ran the tests (updated). I also added the same code to gtk and cocoa, but I cannot test them, so I don't know if they work.